README.markdown
## Redis Master Slave

Redis master-slave client for Ruby, currently only implemented for failover.

## Usage

    require 'redis_master_slave'

    client = RedisMasterSlave.new(YAML.load_file('redis.yml'))

    client.set('a', 1)         # writes to master
    client.get('a')            # reads from slaves, round-robin
    client.master.get('a')     # reads directly from master
    client.slaves[0].get('a')  # reads directly from first slave

### Configuration

The client can be configured in several ways.

#### Single configuration hash

Ideal for configuration via YAML file.

    client = RedisMasterSlave.new(YAML.load_file('redis.yml'))

Example YAML file:

    development:
      master:
          host: localhost
          port: 6379
      slaves:
          - host: localhost
            port: 6380
          - host: localhost
            port: 6381
    production:
      master:
          host: redishost1
          port: 6379
      slaves:
          - host: redishost2
            port: 6380
          - host: redishost3
            port: 6381

#### URL strings

Specify the host and port for each Redis server as a URL string:

    master_url = "redis://localhost:6379"
    slave_urls = [
      "redis://localhost:6380",
      "redis://localhost:6381",
    ]
    client = RedisMasterSlave.new(master_urls, slave_urls)

#### Separate master and slave configurations

Specify master and slave configurations as separate hashes:

    master_config = {:host => 'localhost', :port => 6379}
    slave_configs = []
      {:host => 'localhost', :port => 6380},
      {:host => 'localhost', :port => 6381},
    ]
    client = RedisMasterSlave.new(master_config, slave_configs)

Each configuration hash is passed directly to `Redis.new`.

#### Redis client objects

You can also pass your own Redis client objects:

    master = Redis.new(:host => 'localhost', :port => 6379)
    slave1 = Redis.new(:host => 'localhost', :port => 6380)
    slave2 = Redis.new(:host => 'localhost', :port => 6381)
    client = RedisMasterSlave.new(master, [slave1, slave2])

## Contributing

 * [Bug reports.](https://github.com/oggy/redis_master_slave/issues)
 * [Source.](https://github.com/oggy/redis_master_slave)
 * Patches: Fork on Github, send pull request.
   * Please include tests where practical.
   * Leave the version alone, or bump it in a separate commit.

## Copyright

Copyright (c) George Ogata. See LICENSE for details.


CHANGELOG
== 0.1.8 2012-05-24

 * adding namespace gem and support

== 0.1.7 2012-05-14

 * making type a pass-through method

== 0.1.6 2012-05-03

 * Setting up failover slaves to only connect when failing over instead of initially.

== 0.1.5 2012-05-01

 * Adding transaction support to failover

== 0.1.4 2012-04-26

 * Adding failover because that's what is needed and tests because, well, they're pretty handy.

== 0.1.3 2012-04-24

 * Adding support for multiple databases
 * Adding better respond_to? method that calls Redis.respond_to?
 * Cleaning up code, removing unnecessary functions

== 0.1.2 2012-04-24

 * Making the client actually use the passed config a second time.

== 0.1.1 2012-04-24

 * Making the client actually use the passed config

== 0.0.1 2011-03-10

 * Hi.


test/test_helper.rb
DEBUG=false

require 'rubygems'
require 'test/unit'
require 'mocha'
require 'redis_master_slave'

PORT    = 6379
OPTIONS_HASH = {:port => PORT, :db => 15, :timeout => 3}
OPTIONS_STR   = "redis://127.0.0.1:#{PORT}/15"
NODES   = ["redis://127.0.0.1:6380/15"]

class FakeRedisClient
  attr_accessor :name
  attr_accessor :db
  attr_accessor :call
  attr_accessor :max_delayed_calls
  attr_accessor :in_multi_block
  attr_accessor :queue
  attr_accessor :hash

  def initialize(name=nil)
    @name = name
    @hash = {0=>{}}
    @db = 0
    @max_delayed_calls = 0
    @call = 0
    @in_multi_block=nil
    @queue=nil
  end

  def get(key)
    @hash[@db][key]
  end

  def set(key, val)
    @call+=1
    if (@call<=@max_delayed_calls)
      puts "sleeping #{@call}:#{@max_delayed_calls}" if DEBUG
      sleep(15)
    end
    @hash[@db][key]=val
    "OK"
  end
  
  def select(i)
    @db=i
    @hash[@db]={}
  end

  # TODO: Stub out MULTI and EXEC
end


test/test_failover.rb
require 'test_helper'

class FailoverTest < Test::Unit::TestCase
  def test_master
    rms = RedisMasterSlave.new(FakeRedisClient.new,[FakeRedisClient.new,FakeRedisClient.new])

    assert_equal(nil, rms.get("a"))
    rms.set("a",2)
    assert_equal(2, rms.get("a"))
    rms.set("a",3)
    assert_equal(3, rms.get("a"))
  end

  # This currently requires a real running redis instance.
  def test_master_multi_block
    rms = RedisMasterSlave.new(Redis.new(OPTIONS_HASH))

    assert_equal(rms.multi {rms.set("a",2);rms.get("a")}, 
                 ["OK", "2"])
    assert_equal(rms.get("a"),"2")
    assert_equal(rms.multi do
                   rms.set("b",3);
                   rms.get("b")
                   rms.get("a")
                   rms.set("a",4);
                 end, 
                 ["OK", "3", "2", "OK"])
    assert_equal(rms.get("a"),"4")
  end

  # TODO: add WATCH and UNWATCH tests.
  def test_master_multi
    rms = RedisMasterSlave.new(Redis.new(OPTIONS_HASH))

    # Make sure you can discard the transaction
    assert_equal(rms.set("a",1), "OK")
    assert_equal(rms.multi, "OK")
    assert_equal(rms.set("a",2), "QUEUED")
    assert_equal(rms.get("a"), "QUEUED")
    assert_equal(rms.discard, "OK")
    assert_equal(rms.get("a"), "1")

    # Make sure you can commit the transaction
    assert_equal(rms.multi, "OK")
    assert_equal(rms.set("a",2), "QUEUED")
    assert_equal(rms.get("a"), "QUEUED")
    assert_equal(rms.exec, ["OK", "2"])
    assert_equal(rms.get("a"), "2")

  end
  
  def test_next_failover_slave
    master=FakeRedisClient.new("master")
    slave0=FakeRedisClient.new("slave0")
    slave1=FakeRedisClient.new("slave1")
    rms = RedisMasterSlave.new(master,[slave0,slave1])
    
    assert_equal(slave0, rms.next_failover_slave!)
    assert_equal(slave1, rms.next_failover_slave!)
    assert_nil(rms.next_failover_slave!)
    assert_nil(rms.next_failover_slave!)
  end
  
  def test_manual_failover
    master=FakeRedisClient.new("master")
    slave0=FakeRedisClient.new("slave0")
    slave1=FakeRedisClient.new("slave1")
    rms = RedisMasterSlave.new(master,[slave0,slave1])

    assert_equal("active", rms.status)
    assert_equal(rms.acting_master, master)
    assert_equal(0, rms.failover_index)
    rms.failover!
    assert_equal("failover", rms.status)
    assert_equal(1,rms.failover_index)
    assert_equal(slave0, rms.acting_master)
    rms.failover!
    assert_equal("failover", rms.status)
    assert_equal(slave1, rms.acting_master)
    assert_equal(2,rms.failover_index)
    assert_raise RedisMasterSlave::PermanentFail do 
      rms.failover!
    end
    assert_equal("permanent_fail", rms.status)
    assert_equal(slave1, rms.acting_master)
    assert_equal(3,rms.failover_index)
    assert_raise RedisMasterSlave::PermanentFail do 
      rms.failover!
    end
    assert_equal("permanent_fail", rms.status)
    assert_equal(slave1, rms.acting_master)
    assert_equal(3,rms.failover_index)
  end

  def test_simple_failover
    master=FakeRedisClient.new("master")
    slave0=FakeRedisClient.new("slave0")
    slave1=FakeRedisClient.new("slave1")
    master.max_delayed_calls=10
    rms = RedisMasterSlave.new(master,[slave0,slave1])
    rms.timeout=1
    rms.retry_times=3

    assert_equal("active", rms.status)
    assert_nil(rms.get("a"))
    assert_raise RedisMasterSlave::FailoverEvent do 
      rms.set("a",2)
    end
    assert_equal("failover", rms.status)
    assert_equal(slave0, rms.acting_master)
    assert_equal(2, rms.get("a"))
  end

  def test_simple_failover_with_select
    master=FakeRedisClient.new("master")
    slave0=FakeRedisClient.new("slave0")
    slave1=FakeRedisClient.new("slave1")

    master.max_delayed_calls=10
    rms = RedisMasterSlave.new(master,[slave0,slave1])
    rms.timeout=1
    rms.retry_times=3
    rms.select(2)

    assert_equal("active", rms.status)
    assert_equal(2,rms.db)
    assert_equal(2,rms.acting_master.db)
    assert_nil(rms.get("a"))
    assert_raise RedisMasterSlave::FailoverEvent do 
      rms.set("a",2)
    end
    assert_equal(slave0, rms.acting_master)
    assert_equal(2,rms.acting_master.db)
    assert_equal(2, rms.get("a"))
  end

  def test_timeout_doesnt_failover
    master=FakeRedisClient.new("master")
    slave0=FakeRedisClient.new("slave0")
    slave1=FakeRedisClient.new("slave1")

    rms = RedisMasterSlave.new(master,[slave0,slave1])
    rms.timeout=1
    rms.retry_times=3

    # Testing if it fails once, then recovers.
    assert_equal("active", rms.status)
    master.max_delayed_calls=1
    assert_equal(nil, rms.get("a"))
    rms.set("a",2)
    assert_equal(2, rms.get("a"))
    assert_equal(rms.acting_master, master)
    assert_equal("active", rms.status)

    # Testing if it fails 2 times in a row, then recovers.
    master.call=0
    master.max_delayed_calls=2
    rms.set("a",3)
    assert_equal(3, rms.get("a"))
    assert_equal(rms.acting_master, master)
    assert_equal("active", rms.status)
  end

  def test_multiple_failover 
    puts "test_multiple_failover" if DEBUG

    master=FakeRedisClient.new("master")
    slave0=FakeRedisClient.new("slave0")
    slave1=FakeRedisClient.new("slave1")

    if DEBUG
      puts "master = #{master.inspect}"
      puts "slave0 = #{slave0.inspect}"
      puts "slave1 = #{slave1.inspect}"
    end

    rms = RedisMasterSlave.new(master,[slave0,slave1])

    rms.timeout=1
    rms.retry_times=3
    master.max_delayed_calls=100

    # Make sure it fails over to slave0 first
    assert_equal("active", rms.status)
    assert_equal(nil, rms.get("a"))
    assert_raise RedisMasterSlave::FailoverEvent do 
      rms.set("a",2)
    end

    if DEBUG
      puts "master = #{master.inspect}"
      puts "slave0 = #{slave0.inspect}"
      puts "slave1 = #{slave1.inspect}"
    end

    assert_equal("failover", rms.status)
    assert_equal(rms.acting_master, slave0)
    assert_equal(2, rms.get("a"))

    # Test second failover
    slave0.call=0
    slave0.max_delayed_calls=10

    assert_equal(2, rms.get("a"))
    assert_raise RedisMasterSlave::FailoverEvent do 
      rms.set("a",3)
    end
    assert_equal(3, rms.get("a"))
    assert_equal(rms.acting_master, slave1)
    assert_equal("failover", rms.status)

    # Test third failover doesn't change redis instances
    slave1.call=0
    slave1.max_delayed_calls=15
    slave0.call=0
    slave0.max_delayed_calls=0

    if DEBUG
      puts "master = #{master.inspect}"
      puts "slave0 = #{slave0.inspect}"
      puts "slave1 = #{slave1.inspect}"
    end

    assert_equal(3, rms.get("a"))
    assert_raise RedisMasterSlave::PermanentFail do 
      rms.set("a",4)
    end
    assert_equal(3, rms.get("a"))
    assert_equal("permanent_fail", rms.status)

    slave0.call=0
    slave0.max_delayed_calls=20
    slave1.call=0
    slave1.max_delayed_calls=21
    assert_equal(rms.acting_master, slave1)

    puts "slave1 = #{slave1.inspect}" if DEBUG

    assert_raise RedisMasterSlave::PermanentFail do 
      rms.set("a",5)
    end
    assert_equal(rms.acting_master, slave1)
    assert_equal("permanent_fail", rms.status)
  end    
  
  def test_no_init_slave_connection
    master=FakeRedisClient.new("master")
    slave0=FakeRedisClient.new("slave0")
    slave1=FakeRedisClient.new("slave1")
    rms = RedisMasterSlave.new(master,[slave0,slave1])

    assert_equal("active", rms.status)
    assert_equal(master, rms.acting_master)
    assert_equal({}, rms.failover_slaves)
    rms.failover!
    assert_equal("failover", rms.status)
    assert_equal(slave0, rms.acting_master)
    assert_equal(slave0, rms.failover_slaves[0])
    assert_equal(1, rms.failover_slaves.size)
    rms.failover!
    assert_equal("failover", rms.status)
    assert_equal(slave1, rms.acting_master)
    assert_equal(slave0, rms.failover_slaves[0])
    assert_equal(slave1, rms.failover_slaves[1])
    assert_equal(2, rms.failover_slaves.size)
    assert_raise RedisMasterSlave::PermanentFail do
      rms.failover!
    end
    assert_equal("permanent_fail", rms.status)
    assert_equal(rms.acting_master, slave1)
    assert_equal(slave0, rms.failover_slaves[0])
    assert_equal(slave1, rms.failover_slaves[1])
    assert_equal(3, rms.failover_slaves.size)
  end

  def test_dry_run
    master=FakeRedisClient.new("master")
    slave0=FakeRedisClient.new("slave0")
    slave1=FakeRedisClient.new("slave1")
    rms = RedisMasterSlave.new(master,[slave0,slave1])
    rms.mode="dryrun"

    assert_equal("active", rms.status)
    assert_equal(master, rms.acting_master)
    assert_equal({}, rms.failover_slaves)
    rms.failover!
    assert_equal("active", rms.status)
    assert_equal(master, rms.acting_master)
    assert_equal({}, rms.failover_slaves)
    rms.failover!
    assert_equal("active", rms.status)
    assert_equal(master, rms.acting_master)
    assert_equal({}, rms.failover_slaves)
  end

  def test_dry_run_with_timeout
    master=FakeRedisClient.new("master")
    slave0=FakeRedisClient.new("slave0")
    slave1=FakeRedisClient.new("slave1")
    master.max_delayed_calls=10
    rms = RedisMasterSlave.new(master,[slave0,slave1])
    rms.timeout=1
    rms.retry_times=3
    rms.mode="dry_run"
    rms.select(2)

    assert_nil(rms.get("a"))
    assert_equal("active", rms.status)
    assert_equal(2,rms.current_db)
    assert_equal(2,rms.acting_master.db)
    assert_raise RedisMasterSlave::FailoverEvent do 
      rms.set("a",2)
    end
    assert_equal("active", rms.status)
    assert_equal(master, rms.acting_master)
    assert_nil(rms.get("a"))
    assert_equal(2,rms.current_db)
    assert_equal(2,rms.acting_master.db)
  end

  # # def test_recover_after_failover
  # #   master=FakeRedisClient.new("master")

  # #   rms = RedisMasterSlave.new(master,)

end


test/test_namespace.rb
require 'test_helper'

class Rails
  class Log
    def self.debug(str)
      puts str
    end
  end

  def self.logger
    Log
  end
end

class NamespaceTest < Test::Unit::TestCase
  def test_namespace_nil
    rms = RedisMasterSlave.new(OPTIONS_HASH)
    r=Redis.new(OPTIONS_HASH)

    assert_nil(rms.namespace)
    assert_nil(rms.namespace_rms)
    rms.set("a","2")
    assert_equal("2", rms.get("a"))
    assert_equal("2", r.get("a"))
    rms.set("a","3")
    assert_equal("3", rms.get("a"))
    assert_equal("3", r.get("a"))
  end

  def test_namespace_non_nil
    rms = RedisMasterSlave.new(OPTIONS_HASH.merge({:namespace => "nachos"}))
    r=Redis.new(OPTIONS_HASH)

    assert_equal("nachos", rms.namespace)
    assert_equal("nachos", rms.namespace_rms)
    rms.set("a","2")
    assert_equal("2", rms.get("a"))
    assert_equal("2", r.get("nachos:a"))
    rms.set("a","3")
    assert_equal("3", rms.get("a"))
    assert_equal("3", r.get("nachos:a"))
  end

  def test_namespace_set_runtime
    rms = RedisMasterSlave.new(OPTIONS_HASH)

    assert_nil(rms.namespace)
    assert_nil(rms.namespace_rms)
    rms.namespace = "popcorn"
    assert_equal("popcorn", rms.namespace)
    assert_equal("popcorn", rms.namespace_rms)

    r=Redis.new(OPTIONS_HASH)
    rms.set("a","2")
    assert_equal("2", rms.get("a"))
    assert_equal("2", r.get("popcorn:a"))
    rms.set("a","3")
    assert_equal("3", rms.get("a"))
    assert_equal("3", r.get("popcorn:a"))
  end

  def test_namespace_set_runtime_with_string
    rms = RedisMasterSlave.new(OPTIONS_STR)

    assert_nil(rms.namespace)
    assert_nil(rms.namespace_rms)
    rms.namespace = "pickles"
    assert_equal("pickles", rms.namespace)
    assert_equal("pickles", rms.namespace_rms)

    r=Redis.new(OPTIONS_HASH)
    rms.set("a","2")
    assert_equal("2", rms.get("a"))
    assert_equal("2", r.get("pickles:a"))
    rms.set("a","3")
    assert_equal("3", rms.get("a"))
    assert_equal("3", r.get("pickles:a"))
  end

end


lib/redis_master_slave/client.rb
require 'uri'
require 'timeout'
require 'redis/namespace'

module RedisMasterSlave
  class FailoverEvent < StandardError; end
  class PermanentFail < StandardError; end
  #
  # Wrapper around a pair of Redis connections, one master and one
  # slave.
  #
  # Read requests are directed to the slave, others are sent to the
  # master.
  #
  class Client
    #
    # Create a new client.
    #
    # +master+ and +slave+ may be URL strings, Redis client option
    # hashes, or Redis clients.
    #
    def initialize(*args)
      @master_config, @slave_configs = *args
      raise ArgumentError, "wrong number of arguments (#{args.size} for 1..2)" if args.size > 2

      @acting_master = @master = make_client(@master_config) or
        extend ReadOnly
      @slave_configs ||= {}
      @failover_index    = 0
      @failover_slaves   = {}
      @timeout           = 15
      @retry_times       = 5
      @mode              = "failover"
      @current_db      ||= 0
      @namespace_rms   ||= nil
      @status            = "active"
    end

    #
    # The master client.
    #
    attr_accessor :master

    #
    # The client who is acting as master (normally @master)
    #
    attr_accessor :acting_master

    #
    # The array of slave clients.
    #
    attr_accessor :failover_slaves

    #
    # Index of the slave to use for the next failover.
    #
    attr_accessor :failover_index

    #
    # Amount of time before redis will timeout
    # 
    attr_accessor :timeout

    #
    # Number of times to timeout before redis will failover to slave
    # 
    attr_accessor :retry_times

    #
    # Current Database being used by active_master
    # 
    attr_accessor :current_db

    #
    # Mode for the master_slave to operate: "failover" or "dry_run"
    #
    attr_accessor :mode

    #
    # namespace_rms defined for the Redis::Namespace object.  Default = nil
    #
    attr_accessor :namespace_rms

    #
    # current state of the instance
    #
    attr_accessor :status

    # :type needs to go the the Redis instance
    undef_method :type

    #
    # Select a specific db for all redis masters and slaves
    #
    # TODO: This needs to be refactored with method_missing, as it's the same logic
    #
    def select(db)
      # Rails.logger.debug("RedisMasterSlave:select(#{db}) on acting_master: #{@acting_master}")
      i=0
      remaining_failovers = (@mode=="failover") ? @slave_configs.size - @failover_index : 0
      @failover_ctr=0

      begin
        Timeout.timeout(@timeout) do
          @acting_master.select(db)
        end
        @current_db=db
      rescue Timeout::Error, Errno::ECONNREFUSED => e
        # Rails.logger.debug("RedisMasterSlave: Error caught in select(#{db}) (i:#{i}; j:#{remaining_failovers})")
        if (i+=1)>=@retry_times
          i=0
          remaining_failovers-=1
          @failover_ctr+=1
          failover!(true)
        end
        
        # make sure we only failover if there's a slave
        if (remaining_failovers>=0)
          # Rails.logger.debug("RedisMasterSlave: retrying (i:#{i}; j:#{remaining_failovers})")
          retry
        end
      ensure
        raise RedisMasterSlave::FailoverEvent if @failover_ctr>0
      end
    end

    #
    # Failover to the next slave
    # 
    def failover!(skip_select = false)
      # Rails.logger.debug("RedisMasterSlave: failover!")
      if @mode=="failover"
        next_slave = next_failover_slave!
        @status = "failover"
        unless next_slave
          # Leaving this > 0 will trump the PermanentFail error with a Failover Event
          @failover_ctr=0
          @status = "permanent_fail"
          raise RedisMasterSlave::PermanentFail 
        end
        @acting_master = next_slave
        # Make sure to stay on same db as old master.
        @acting_master.select(@current_db) unless skip_select
        # raise RedisMasterSlave::FailoverEvent
      end
    end

    #
    # Return the next failover slave to use.  Initialized the redis instance if necessary.
    #
    def next_failover_slave!
      # Rails.logger.debug("RedisMasterSlave: next_failover_slave!: failover_index: #{@failover_index}")
      @failover_index=[@failover_index, @slave_configs.size].min + 1
      @failover_slaves[@failover_index-1] ||= make_client(@slave_configs[@failover_index-1])
    end

    # Specifically for transactions.  EXEC, DISCARD, UNWATCH and WATCH are handled as normal.
    def multi
      if !block_given?
        @acting_master.multi
      else
        @acting_master.multi{yield}
      end
    end

    def namespace=(ns=nil)
      @namespace_rms = ns
      @acting_master.namespace=ns
    end

    # 
    # TODO: make the method_missing memoize a class method
    # TODO: This needs to be refactored with seled method, as it's the same logic
    # 
    def method_missing(method, *params, &block) # :nodoc:
      # Rails.logger.debug("RedisMasterSlave:#{method}(#{params*', '})")
      if @acting_master.respond_to?(method)

        i=0
        remaining_failovers = (@mode=="failover") ? [@slave_configs.size - @failover_index,0].max : 0
        @failover_ctr=0

        begin
          Timeout.timeout(@timeout) do
            @acting_master.send(method, *params, &block)
          end

        rescue Timeout::Error, Errno::ECONNREFUSED
          # Rails.logger.debug("RedisMasterSlave: Error Caught in call #{method}(#{params}) (i:#{i}; j:#{j})")
          if (i+=1)>=@retry_times
            i=0
            remaining_failovers-=1
            @failover_ctr+=1
            failover!
          end
          
          # make sure we only failover if there's a slave
          if (remaining_failovers>=0)
            # Rails.logger.debug("RedisMasterSlave: retrying (i:#{i}; j:#{remaining_failovers})")
            retry
          end
        ensure
          raise RedisMasterSlave::FailoverEvent if @failover_ctr>0
        end
      else
        super
      end
    end

    # helpful for testing
    def throw_failover_event_exception
      raise RedisMasterSlave::FailoverEvent
    end

    def throw_general_exception
      raise Exception
    end

    # Note that this drops the include_private parameter, as Namespace doesn't support it.
    def respond_to_with_redis?(symbol, include_private=false)
      respond_to_without_redis?(symbol, include_private) || 
        @acting_master.respond_to?(symbol)
    end
    alias_method :respond_to_without_redis?, :respond_to?
    alias_method :respond_to?, :respond_to_with_redis?

    private

    def make_client(config)
      # Rails.logger.debug("RedisMasterSlave: make_client(#{config.inspect})")
      case config
      when String
        # URL like redis://localhost:6379/db
        uri = URI.parse(config)
        db = uri.path.split("/")[1].to_i
        db ||= 0

        r = Redis.new(:host => uri.host, :port => uri.port, :db => db)
        Redis::Namespace.new(@namespace_rms, :redis => r)
      when Hash
        # Hash of Redis client options (string keys ok).
        redis_config = {}
        config.each do |key, value|
          redis_config[key.to_sym] = value
        end
        r = Redis.new(redis_config)
        @namespace_rms = config[:namespace]
        Redis::Namespace.new(@namespace_rms, :redis => r)
      else
        # Hopefully a client object.
        config
      end
    end
  end
end
